Core: Move `react` and `react-dom` to peer deps

gaetanmaisse opened the pull request:
Follow up https://github.com/storybookjs/storybook/pull/12908#issuecomment-717261800
More or less revert https://github.com/storybookjs/storybook/pull/11628

## What I did

As stated by @Hypnosphi:
> `react` and `react-dom` should always be peer dependencies of packages that provide react components to ensure that there's a single instance of those.

So I moved them to peerDep in some base packages. Plus, all packages depending on a package with `react` as peerDep should also have it as peerDeps (transitive peer dependencies, for details: https://dev.to/arcanis/implicit-transitive-peer-dependencies-ed0).

Finally, packages in `examples/*` are "end" packages using Storybook so they must provide `react` and `react-dom` in their deps/devDeps. So, I put `react` and `react-dom` to `^16.14.0` (latest 16.x available).

## How to test

- CI should be 🟢 

## Questions
 - I used `15 || 16 || ^17.0.0` as version range for both `react` and `react-dom` when they are peerDep. Does it look good to you?
 - Some `app` packages should maybe have `react` and `react-dom` as deps instead of peerDeps, I mean `@storybook/angular` would find weird to have to add these deps in their project themselves. 🤷🏻 
 - If we keep `react` and `react-dom` as peerDeps in `app` packages we should update the CLI to detect and add if needed add them automagically when running `sb init`
 - Should we add an example with React 17? We now have one for React 16 and one for React 15

Did I miss something @Hypnosphi? and what's your opinion on that?

gaetanmaisse committed:
Move `react` and `react-dom` to peer deps

`react` and `react-dom` should always be peer dependencies of packages that provide react components to ensure that there's a single instance of those.

Plus, all packages depending on a package with `react` as peerDep should also have it as peerDeps (transitive peer dependencies) or as deps.

Finally, packages in `examples/*` are "end" packages using Storybook so they must provide `react` and `react-dom` in their deps/devDeps

Hypnosphi commented:
I'd just restore the state before #11628, it more or less worked that way

gaetanmaisse committed:
Use `react` and `react-dom` 16.13.1 as 16.14.0 looks to cause trouble with examples/official-storybook

Example:
ERR! SyntaxError: /tmp/storybook/examples/official-storybook/stories/notes/notes.md: pragma and pragmaFrag cannot be set when runtime is automatic.

gaetanmaisse committed:
Add `react` and `react-dom` to Preact and React TS E2E tests

These deps need to be explicitly added as they are peerDeps of SB packages now.

shilman commented on addons/docs/package.json:
```diff
@@ -120,6 +118,8 @@
     "@babel/core": "^7.11.5",
     "@storybook/vue": "6.1.0-alpha.34",
     "babel-loader": "^8.0.0",
+    "react": "15 || 16 || ^17.0.0",
```
This is wrong. Addon-docs requires hooks which require 16.8.

gaetanmaisse replied:
Do you have in mind other packages requiring React 16.8+? Do you prefer I use `^16.8.3 || ^17.0.0` everywhere to be more confident about the compatibility?

shilman replied:
Yes I think that's safe

gaetanmaisse committed:
Update `react` and `react-dom` range version in some addons using React Hooks

gaetanmaisse committed:
Switch `react` and `react-dom` to regular dep in non-react apps

I used `16.13.1` as it looks there are, for now, some issue between 16.14.x and `addons-docs`

gaetanmaisse commented:
Notes: Use `peerDependenciesMeta` to make `react` and `react-dom` optional in all addons? And so avoid users to have the Yarn/npm warning

Hypnosphi commented:
> Use `peerDependenciesMeta` to make react and react-dom optional in all addons?

Addons won't work without react, so it's not optional

ndelangen commented:
> Addons won't work without react, so it's not optional

But we assure it's there anyway!?

gaetanmaisse commented:
You're right @Hypnosphi it's just a workaround we discuss this morning to avoid yarn/npm warnings.

Do you have something in mind for this kind of feedbacks:
> As a dev on an angular/vue project I don't want to have to add `react` and `react-dom` to my project as it's Storybook internal stuff. Storybook should work out of the box with `@storybook/angular(/vue)` and some addons.

For the `@storybook/[app]` (non react apps) we can add React as a regular dep because there will be no React version conflict. But what can we do for addons 🤔  They need react but we are sure to have it because addons cannot work without a `@storybook/[app]`, that's why we come up with this optional peerDep workaround 🤯 

Other options: we keep react as peerDep everywhere and users need to add react to their own project. But @ndelangen told me it can be a pain to deal with react in a vue project for instance. 🤷🏻

Hypnosphi commented:
OK let's make them optional until NPM 7 becomes mainstream (probably in a year)

shilman commented:
In https://github.com/storybookjs/storybook/pull/12925 I made them peer deps for everything except the non-React frameworks. So `@storybook/vue` will install React internally. I think this is the right split.

gaetanmaisse commented:
Monthly meeting sum up:
https://github.com/storybookjs/community/blob/master/meetings/2020/monthly-2020-10-02.md#react-as-peerdep-vs-dep

Still one pending question:
 - A fixed version of `react` in all examples?

shilman commented:
@gaetanmaisse 

> A fixed version of react in all examples?

Yes except `cra-react15`, unless we want to move that to `examples-v2`

gaetanmaisse committed:
Merge branch 'next' into tech/move-react-and-react-dom-to-peer-deps

gaetanmaisse committed:
Apply solution discussed in monthly meeting:

- Set `react` and `react-dom` as peerDeps for all `lib` packages needing them
- Set `react` and `react-dom` as peerDeps + optional tag for all `addon` packages needing them
- Set `react` and `react-dom` as regular deps for all non react apps
- Remove `react` and `react-dom` from root `package.json`
- Set a fixed version of `react` in all examples

Also:
- Use "^16.8.0 || ^17.0.0" version range as some packages need React Hooks

gaetanmaisse committed:
Remove `react` and `react-dom` from required deps in E2E test configs

They are now regular deps of SB packages for all non-react app so there is no more need to add them explicitly.

tmeasday commented:
Shouldn't our (non-react) examples only install react via the `@storybook/X` framework package? Or does that lead to peer dep warnings and we don't like that?

shilman committed:
Update MDX snapshots

shilman committed:
Update yarn.lock

shilman reviewed (commented):
Looking great @gaetanmaisse

shilman commented:
@tmeasday this PR already implements that

stof commented:
> this PR already implements that (non-react frameworks automatically install react on behalf of users)

that's actually not true for package manager which are strict about peer dependencies (pnpm and yarn 2 for instance). The `react` dependency of `@storybook/html` will satisfy a peer dependency of a package being a dependency of `@storybook/html`, but not of an addon being a dependency of the project and so a sibling of `@storybook/html`.
And if the project adds a `react` requirement to satisfy the peer dependency, you end up again with 2 different versions of react.

gaetanmaisse commented:
@stof Actually it's still a bit more complex than that because there is a webpack alias in SB manager config so all React imports are aliased to the React version that satisfies `@storybook/core` peer dep. So even if there is another version of React to satisfy an addon peer dep it won't be used by Storybook. 
I have created https://github.com/storybookjs/storybook/issues/13194 + comments to sum-up the react dep situation, feel free to add a comment if something is unclear!

stof commented:
@gaetanmaisse thanks for the link
